### PR TITLE
[275] Implement ExposeService method

### PR DESCRIPTION
**Pull request by jiripetrlik**
# Issue link
Issue: https://github.com/project-codeflare/codeflare-operator/issues/275

PR to distributed-workloads with changes to tests will follow.

### CODE REVIEW

File: test/support/route.go [hunk 1]
```diff
@@ -36,3 +40,45 @@ func GetRoute(t Test, namespace, name string) *routev1.Route {
...
+	// Retrieve hostname
+	r, err = t.Client().Route().RouteV1().Routes(r.Namespace).Get(t.Ctx(), r.Name, metav1.GetOptions{})
+	t.Expect(err).NotTo(gomega.HaveOccurred())
+	hostname := r.Status.Ingress[0].Host
+
+	// Wait for expected HTTP code
+	t.Eventually(func() int {
+		resp, _ := http.Get("http://" + hostname)
+		return resp.StatusCode
+	}, TestTimeoutLong).Should(gomega.Equal(expetedCode))
```
Comment (sutaakar) [inline test/support/route.go:81]: What about waiting until Route stops returning HTTP 503 and a text in response body referring to the Pods being down?
That would allow dropping expected code parameter.

File: test/support/route.go [hunk 2]
```diff
@@ -36,3 +40,45 @@ func GetRoute(t Test, namespace, name string) *routev1.Route {
 	t.T().Helper()
 	return Route(t, namespace, name)(t)
 }
+
+func ExposeService(t Test, namespace string, service string, route string, port string, expetedCode int) *routev1.Route {
```
Comment (sutaakar) [inline test/support/route.go:44]: The parameter names may be misleading, what about:
serviceName
servicePort
routeName

**Comment by jiripetrlik** [comment]
@sutaakar 
Thank you for the review. I've tried to address both your comments.

File: test/support/route.go [hunk 3]
```diff
@@ -36,3 +40,45 @@ func GetRoute(t Test, namespace, name string) *routev1.Route {
 	t.T().Helper()
 	return Route(t, namespace, name)(t)
 }
+
+func ExposeService(t Test, namespace string, serviceName string, routeName string, servicePort string) *routev1.Route {
```
Comment (sutaakar) [inline test/support/route.go:44]: A nitpick: 
```suggestion
func ExposeService(t Test, namespace string, serviceName, servicePort, routeName string) *routev1.Route {
```
To keep service related properties together

File: test/support/route.go [hunk 4]
```diff
@@ -36,3 +40,45 @@ func GetRoute(t Test, namespace, name string) *routev1.Route {
...
+	t.Eventually(Route(t, r.Namespace, r.Name), TestTimeoutLong).
+		Should(gomega.WithTransform(ConditionStatus(routev1.RouteAdmitted), gomega.Equal(corev1.ConditionTrue)))
+
+	// Retrieve hostname
+	r, err = t.Client().Route().RouteV1().Routes(r.Namespace).Get(t.Ctx(), r.Name, metav1.GetOptions{})
+	t.Expect(err).NotTo(gomega.HaveOccurred())
+	hostname := r.Status.Ingress[0].Host
+
+	// Wait for expected HTTP code
+	t.Eventually(func() int {
```
Comment (sutaakar) [inline test/support/route.go:78]: Can you also check that response body contains text specific to the Route not serving yet, for example:
`Route and path matches, but all pods are down.`

File: test/support/route.go (hunk 3 above)
Comment (sutaakar) [reply test/support/route.go:44]: Also makes me think, do we need to specify route name? Is there any case when caller would need this info?
What about using service name as route name, possibly with some suffix? That way we reasonably assure that Route name is unique.

File: test/support/route.go (hunk 4 above)
Comment (sutaakar) [reply test/support/route.go:78]: The reason is that theoretically the 503 can be a correct response code for the service.
Though that may be really edge case.

File: test/support/ingress.go [hunk 5]
```diff
@@ -39,3 +39,43 @@ func GetIngress(t Test, namespace, name string) *networkingv1.Ingress {
 func LoadBalancerIngresses(ingress *networkingv1.Ingress) []networkingv1.IngressLoadBalancerIngress {
 	return ingress.Status.LoadBalancer.Ingress
 }
+
+func exposeServiceIngress(t Test, ingressName string, namespace string, serviceName string, servicePort string) *networkingv1.Ingress {
+	prefixType := networkingv1.PathTypePrefix
+	ingress := &networkingv1.Ingress{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: networkingv1.SchemeGroupVersion.String(),
+			Kind:       "Route",
```
Comment (sutaakar) [inline test/support/ingress.go:48]:
```suggestion
			Kind:       "Ingress",
```

File: test/support/test_parameters.go [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+package support
+
+import (
+	"os"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
```
Comment (sutaakar) [inline test/support/test_parameters.go:26]: My idea was to have here just constants, not variables.
Something like https://github.com/opendatahub-io/distributed-workloads/blob/8b189dc847fe56f7bf57d69c002a5d85defe3c40/tests/integration/config.go

File: test/support/service.go [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+package support
+
+func ExposeService(t Test, name string, namespace string, serviceName string, servicePort string) string {
+	if IsOpenShift(t) {
+		route := ExposeServiceRoute(t, namespace, serviceName, name, servicePort)
+		return "http://" + route.Status.Ingress[0].Host
+	} else {
+		exposeServiceIngress(t, namespace, serviceName, name, servicePort)
+		return KubernetesHostname + "/" + name
```
Comment (sutaakar) [inline test/support/service.go:25]:
```suggestion
		return "http://" + KubernetesHostname + "/" + name
```

File: test/support/route.go (hunk 4 above)
Comment (sutaakar) [reply test/support/route.go:78]: Currently not needed.

File: test/support/ingress.go [hunk 8]
```diff
@@ -39,3 +39,43 @@ func GetIngress(t Test, namespace, name string) *networkingv1.Ingress {
...
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
```
Comment (sutaakar) [inline test/support/ingress.go:79]: Can you create the Ingress on cluster and make sure it is ready? Same like it is done in Route.

**Comment by sutaakar** [comment]
Please use the new function in https://github.com/project-codeflare/codeflare-operator/blob/a14914da37d61dd2e3de160c7033e3b2fe6d500e/test/e2e/mnist_rayjob_mcad_raycluster_test.go to make sure it works properly.
Also the `KUBERNETES_HOSTNAME` needs to be set in PR check workflow.

**Comment by jiripetrlik** [comment]
Hello @sutaakar , I've tried to address all the comments. In the final version I removed KUBERNETES_HOSTNAME parameter at all.

File: test/support/service.go [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+*/
+
+package support
+
+import "net/url"
+
+func ExposeService(t Test, name string, namespace string, serviceName string, servicePort string) url.URL {
+	if IsOpenShift(t) {
+		route := ExposeServiceRoute(t, name, namespace, serviceName, servicePort)
+		route = GetRoute(t, route.Namespace, route.Name)
```
Comment (sutaakar) [inline test/support/service.go:24]: This can be moved to `ExposeServiceRoute`, same as done in Ingress.

File: test/support/ingress.go [hunk 10]
```diff
@@ -39,3 +39,56 @@ func GetIngress(t Test, namespace, name string) *networkingv1.Ingress {
 func LoadBalancerIngresses(ingress *networkingv1.Ingress) []networkingv1.IngressLoadBalancerIngress {
 	return ingress.Status.LoadBalancer.Ingress
 }
+
+func ExposeServiceIngress(t Test, name string, namespace string, serviceName string, servicePort string) *networkingv1.Ingress {
```
Comment (sutaakar) [inline test/support/ingress.go:43]: Optional suggestion - what about using name `ExposeServiceByIngress`

File: test/support/route.go [hunk 11]
```diff
@@ -36,3 +40,45 @@ func GetRoute(t Test, namespace, name string) *routev1.Route {
 	t.T().Helper()
 	return Route(t, namespace, name)(t)
 }
+
+func ExposeServiceRoute(t Test, name string, namespace string, serviceName string, servicePort string) *routev1.Route {
```
Comment (sutaakar) [inline test/support/route.go:44]: Same here.

File: test/support/service.go [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+
+		serviceURL := url.URL{
+			Scheme: "http",
```
Comment (sutaakar) [inline test/support/service.go:27]: A nitpick, can you derive using HTTP or HTTPS based on Route spec or status content?

File: test/support/route.go [hunk 13]
```diff
@@ -36,3 +41,57 @@ func GetRoute(t Test, namespace, name string) *routev1.Route {
...
+		resp, _ := http.Get("http://" + hostname)
+		return resp.StatusCode
+	}, TestTimeoutLong).Should(gomega.Not(gomega.Equal(503)))
+
+	r = GetRoute(t, r.Namespace, r.Name)
+
+	scheme := "http"
+	if r.Spec.Port.TargetPort.String() == "https" {
+		scheme = "https"
+	}
```
Comment (sutaakar) [inline test/support/route.go:89]: When I think about it again, this is not actually needed, as above we construct HTTP route anyway, so we are sure it is HTTP.
Can you please remove this code?

**Review by sutaakar — APPROVED** [review]
/lgtm
